api/proto/teleport/machineid/v1/bot_instance.proto

```diff
@@ -0,0 +1,116 @@
+  // not be modified by users.
+  BotInstanceStatus status = 6;
+}
+
+// BotInstanceSpec contains fields
+message BotInstanceSpec {
+  // Empty as is not user configurable.
+  // Eventually this could be leveraged for simple command and control?
+}
+
+// BotInstanceStatusHeartbeat contains information self-reported by an instance
+// of a Bot. This information is not verified by the server and should not be
+// trusted.
+message BotInstanceStatusHeartbeat {
+  // The timestamp that the heartbeat was recorded by the Auth Server. Any
+  // value submitted by `tbot` for this field will be ignored.
+  google.protobuf.Timestamp recorded_at = 1;
+  // Indicates whether this is the heartbeat submitted by `tbot` on startup.
+  bool is_startup = 2;
+  // The version of `tbot` that submitted this heartbeat.
+  string version = 3;
+  // The hostname of the host that `tbot` is running on.
+  string hostname = 4;
+  // The duration that `tbot` has been running for when it submitted this
+  // heartbeat.
+  google.protobuf.Duration uptime = 5;
+  // The currently configured join_method.
+  string join_method = 6;
+  // Indicates whether `tbot` is running in one-shot mode.
+  bool one_shot = 7;
+  // In future iterations, additional information can be submitted here.
+  // For example, the configuration of `tbot` or the health of individual
+  // outputs.
+}
+
+// BotInstanceStatusAuthentication contains information about a join or renewal.
+// Ths information is entirely sourced by the Auth Server and can be trusted.
+message BotInstanceStatusAuthentication {
+  // The timestamp that the join or renewal was authenticated by the Auth
+  // Server.
+  google.protobuf.Timestamp authenticated_at = 1;
+  // The join method used for this join or renewal.
+  string join_method = 2;
+  // The join token used for this join or renewal. This is only populated for
+  // delegated join methods as the value for `token` join methods is sensitive.
+  string join_token = 3;
+  // The metadata sourced from the join method.
+  google.protobuf.Struct metadata = 4;
+  // On each renewal, this generation is incremented. For delegated join
+  // methods, this counter is not checked during renewal. For the `token` join
+  // method, this counter is checked during renewal and the Bot is locked out if
+  // the counter in the certificate does not match the counter of the last
+  // authentication.
+  int32 generation = 5;
+  // The public key of the Bot instance.
+  // When authenticating a Bot instance, the full public key must be compared
+  // rather than just the fingerprint to mitigate pre-image attacks.
+  bytes public_key = 6;
+  // The fingerprint of the public key of the Bot instance.
+  string fingerprint = 7;
+}
+
+// BotInstanceStatus holds the status of a BotInstance.
+message BotInstanceStatus {
+  // The unique identifier for this bot.
+  string id = 1;
+  // The name of the Bot that this instance is associated with.
+  string bot_name = 2;
```

Comment:
Yeah - I think the plan is that BotInstance.Metadata.Name will either be `id` or `bot_name/id`. BotInstance's belong to a Bot, which is what's being indicated by `bot_name` here. We plan to store these in the backend with a key of `bot_name/id` so that we can easily find all BotInstances that belong to a bot.

Reply:
Yes, I think the final `.metadata.name` of a bot instance will have to be `$botName/$instanceID`, leading to some tctl UX along the lines of `tctl get botinstance/foo/abcd1234` (I think there's precedent for that?)

Reply:
I think I was off before and the `.metadata.name` will just be the instance ID. This is using the built in service prefixing functionality so the bot name is already included in the key and we don't want to include it twice.